circuit/types/field/src/helpers/to_bits.rs

```diff
@@ -33,44 +33,71 @@ impl<E: Environment> ToBits for Field<E> {
 impl<E: Environment> ToBits for &Field<E> {
     type Boolean = Boolean<E>;
 
-    /// Outputs the little-endian bit representation of `self` *without* trailing zeros.
+    /// Outputs the unique, minimal little-endian bit representation of `self` *without* trailing zeros.
     fn to_bits_le(&self) -> Vec<Self::Boolean> {
         self.bits_le
             .get_or_init(|| {
-                // Construct a vector of `Boolean`s comprising the bits of the field value.
-                let bits_le = witness!(|self| self.to_bits_le());
-
-                // Reconstruct the bits as a linear combination representing the original field value.
-                let mut accumulator = Field::zero();
-                let mut coefficient = Field::one();
-                for bit in &bits_le {
-                    accumulator += Field::from_boolean(bit) * &coefficient;
-                    coefficient = coefficient.double();
+                // Extract a non-unique little-endian bit representation of `self`.
+                let bits_le = self.to_non_unique_bits_le();
+                // Check that the non-unique bit representation is less than the field modulus.
+                {
+                    // Retrieve the modulus & subtract by 1 as we'll check `bits_le` is less than or *equal* to this value.
+                    // (For advanced users) BaseField::MODULUS - 1 is equivalent to -1 in the field.
+                    let modulus_minus_one = -E::BaseField::one();
+
+                    // Compute `!((BaseField::MODULUS - 1) < bits_le)`, which is equivalent to `((BaseField::MODULUS - 1) >= bits_le)`, which is equivalent to `bits_le < BaseField::MODULUS`.
+                    let is_less_than_modulus = !modulus_minus_one.to_bits_le().iter().zip_eq(bits_le.iter()).fold(
+                        Boolean::constant(false),
+                        |rest_is_less, (this, that)| {
+                            if *this { that.bitand(&rest_is_less) } else { that.bitor(&rest_is_less) }
+                        },
+                    );
+
+                    // Ensure that the non-unique bit-representation is less than `BaseField::MODULUS`.
+                    E::assert(is_less_than_modulus);
                 }
 
-                // Ensure value * 1 == (2^i * b_i + ... + 2^0 * b_0)
-                E::assert_eq(*self, accumulator);
-
                 bits_le
             })
             .clone()
     }
 
-    /// Outputs the big-endian bit representation of `self` *without* leading zeros.
+    /// Outputs the unique, minimal big-endian bit representation of `self` *without* leading zeros.
     fn to_bits_be(&self) -> Vec<Self::Boolean> {
         let mut bits_le = self.to_bits_le();
         bits_le.reverse();
         bits_le
     }
 }
 
+impl<E: Environment> Field<E> {
+    /// Outputs a non-unique little-endian bit representation of `self` *without* trailing zeros.
+    fn to_non_unique_bits_le(&self) -> Vec<Boolean<E>> {
```

Comment:
Factoring this into its own function is nice. It mirrors one of the gadgets we have defined in ACL2.